lib/autoupdate/tools/updater.go

```diff
@@ -0,0 +1,391 @@
+	hash, err := u.downloadHash(ctx, pkg.Hash)
+	if pkg.Optional && trace.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	archivePath, archiveHash, err := u.downloadArchive(ctx, u.toolsDir, pkg.Archive)
+	if pkg.Optional && trace.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer func() {
+		if err := os.Remove(archivePath); err != nil {
+			slog.WarnContext(ctx, "failed to remove archive", "error", err)
+		}
+	}()
+	if archiveHash != hash {
+		return trace.BadParameter("hash of archive does not match downloaded archive")
+	}
+
+	pkgName := fmt.Sprint(uuid.New().String(), updatePackageSuffix)
+	extractDir := filepath.Join(u.toolsDir, pkgName)
+	if runtime.GOOS != constants.DarwinOS {
+		if err := os.Mkdir(extractDir, 0o755); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	// Perform atomic replace so concurrent exec do not fail.
+	if err := packaging.ReplaceToolsBinaries(u.toolsDir, archivePath, extractDir, u.tools); err != nil {
+		return trace.Wrap(err)
+	}
+	// Cleanup the tools directory with previously downloaded and un-archived versions.
+	if err := packaging.RemoveWithSuffix(u.toolsDir, updatePackageSuffix, pkgName); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// Exec re-executes tool command with same arguments and environ variables.
+func (u *Updater) Exec() (int, error) {
+	path, err := toolName(u.toolsDir)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+	// To prevent re-execution loop we have to disable update logic for re-execution.
+	env := append(os.Environ(), teleportToolsVersionEnv+"=off")
+
+	if runtime.GOOS == constants.WindowsOS {
+		cmd := exec.Command(path, os.Args[1:]...)
+		cmd.Env = env
+		cmd.Stdin = os.Stdin
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+		if err := cmd.Run(); err != nil {
+			return 0, trace.Wrap(err)
+		}
+
+		return cmd.ProcessState.ExitCode(), nil
+	}
+
+	if err := syscall.Exec(path, append([]string{path}, os.Args[1:]...), env); err != nil {
+		return 0, trace.Wrap(err)
+	}
+
+	return 0, nil
+}
+
+func (u *Updater) downloadHash(ctx context.Context, url string) (string, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	resp, err := u.client.Do(req)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode == http.StatusNotFound {
+		return "", trace.NotFound("hash file is not found: %v", resp.StatusCode)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return "", trace.BadParameter("bad status when downloading archive hash: %v", resp.StatusCode)
+	}
+
+	var buf bytes.Buffer
+	_, err = io.CopyN(&buf, resp.Body, sha256.Size*2) // SHA bytes to hex
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	raw := buf.String()
+	if _, err = hex.DecodeString(raw); err != nil {
+		return "", trace.Wrap(err)
+	}
+	return raw, nil
```

Comment:
Suggest returning the decoded hash and comparing the actual bytes, instead of comparing hex (which has multiple valid representations)